vshard/replicaset.lua

```diff
@@ -75,6 +75,108 @@ local fiber_cond_wait = util.fiber_cond_wait
 local future_wait = util.future_wait
 local gsc = util.generate_self_checker
 
+--
+-- vconnect is asynchronous vshard greeting, saved inside netbox connection.
+-- It stores future object and additional info, needed for its work.
+-- Future is initialized, when the connection is established (inside
+-- netbox_on_connect). The connection cannot be considered "connected"
+-- until vconnect is properly validated.
+--
+local function conn_set_vconnect(conn)
+    assert(conn.replica ~= nil)
+    conn.vconnect = {
+        -- Whether the connection is done to the named replica.
+        is_named = conn.replica.id == conn.replica.name,
+        -- Used to wait for the appearance of vconnect.future object,
+        -- if call is done before the connection is established.
+        future_cond = fiber.cond(),
+    }
+    local vconn = conn.vconnect
+    if not vconn.is_named then
+        -- Skip validation without any waiting.
+        conn.vconnect.is_validated = true
+    end
+end
+
+--
+-- Initialize future object. Should be done, when connection is established
+-- (inside netbox_on_connect).
+--
+local function conn_init_vconnect(conn)
+    local vconn = conn.vconnect
+    if vconn and vconn.is_named then
+        -- Validation should be done after every reconnect.
+        vconn.is_validated = false
+        local opts = {is_async = true}
+        vconn.future = conn:call('vshard.storage._call', {'info'}, opts)
+        vconn.future_cond:broadcast()
+    end
+end
+
+--
+-- Check, that future is ready, and its result is expected, close the connection
+-- if the result is not correct. The function doesn't yield.
+-- @retval true; The correct response is received.
+-- @retval nil, ...; Response is not received or validation error happened.
+--
+local function conn_validate_vconnect(conn)
+    local vconn = conn.vconnect
+    -- conn.vconnect may be nil, if connection was created on old version
+    -- and the storage was reloaded to a new one.
+    if not vconn or vconn.is_validated then
+        return true
+    end
+    -- Nothing to do, but wait in such case.
+    if not vconn.future or not vconn.future:is_ready() then
+        return nil, lerror.vshard(lerror.code.CONNECTION_IS_NOT_ESTABLISHED,
+                                  conn.replica.id)
+    end
+    -- Critical errors. Connection should be closed after these ones.
+    local result, err = vconn.future:result()
+    if not result then
+        conn:close()
+        -- Failed to get response. E.g. access error.
```

Comment:
Can you add a test on that? Should be easy if you just do this `vshard.storage._call = nil` on the storage side.

Reply:
Added tests for all cases, you asked

Reply:
Are you sure? I don't see any new tests.

Reply:
Oh, sorry. It remained in untracked files and wasn't added to commit